File: aoc2015/day03/src/test/java/de/havox_design/aoc2015/day03/Day03Test.java
package de.havox_design.aoc2015.day03;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

class Day03Test {

    @SuppressWarnings("java:S2699")
    @Test
    void testMainClass() {
        MainClass.main(new String[0]);
    }

    @ParameterizedTest
    @MethodSource("getDataForPart1")
    void testPart1(String fileName, int expectedHousesVisited) {
        Assertions.assertEquals(expectedHousesVisited, PresentDelivery.solvePart1(fileName));
    }

    private static Stream<Arguments> getDataForPart1() {
        return Stream.of(
                Arguments.of("day03Part1sample1.txt", 2),
                Arguments.of("day03Part1sample2.txt", 4),
                Arguments.of("day03Part1sample3.txt", 2)
        );
    }

    @ParameterizedTest
    @MethodSource("getDataForPart2")
    void testPart2(String fileName, int expectedHousesVisited) {
        Assertions.assertEquals(expectedHousesVisited, PresentDelivery.solvePart2(fileName));
    }

    private static Stream<Arguments> getDataForPart2() {
        return Stream.of(
                Arguments.of("day03Part2sample1.txt", 3),
                Arguments.of("day03Part2sample2.txt", 3),
                Arguments.of("day03Part2sample3.txt", 11)
        );
    }
}

File: aoc2023/src/test/kotlin/de/havox_design/aoc2023/day03/Day03Test.kt
package de.havox_design.aoc2023.day03

import de.havox_design.aoc2023.MainClass
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream

class Day03Test {
    @Test
    fun testMainClass() {
        MainClass.main(arrayOf("day03"))
    }

    @ParameterizedTest
    @MethodSource("getDataForTestSolvePart1")
    fun testSolvePart1(filename: String, expectedResult: Long) =
            GearRatios(filename).solvePart1().shouldBe(expectedResult)

    @ParameterizedTest
    @MethodSource("getDataForTestSolvePart2")
    fun testSolvePart2(filename: String, expectedResult: Long) =
            GearRatios(filename).solvePart2().shouldBe(expectedResult)

    companion object {
        @JvmStatic
        private fun getDataForTestSolvePart1(): Stream<Arguments> =
                Stream.of(
                        Arguments.of("de/havox_design/aoc2023/day03/day03Sample.txt", 4361L)
                )

        @JvmStatic
        private fun getDataForTestSolvePart2(): Stream<Arguments> =
                Stream.of(
                        Arguments.of("de/havox_design/aoc2023/day03/day03Sample.txt", 467835L)
                )
    }
}

private fun Long.shouldBe(expectation: Long) = assertEquals(expectation, this)
